### PR TITLE
[DataGrid] Refactor edit events to allow stop propagation

### PR DESCRIPTION
**Pull request by dtassone**
Split of #1287 
just the fixes.
Doc will come in a separate PR

### CODE REVIEW

**Review by oliviertassinari — COMMENTED** [review]
I'm not a big fan of using `stopPropagation()` for the purpose of disabling the native feature of the grid. But happy to give it a try. I see 3 downsides:

1. It's not intuitive (to me). It works because React has introduced a nonstandard `isPropagationStopped ` method. Intuitively, I was expecting the usage of `event.stopImmediatePropagation()`.
2. It only works with React. Who knows, maybe in the future we will replicate https://fullcalendar.io/ approach for supporting other frameworks, React first, then Preact wrapped inside Vue.
3. IMHO, in 6 or 12 months from now, we will have developers complaining about how they want to disable the native action of the grid and complain because the event doesn't bubble anymore. In which case, we would need to use this approach https://next.material-ui.com/components/autocomplete/#events.

File: packages/grid/_modules_/grid/components/containers/GridRootStyles.ts [hunk 1]
```diff
@@ -129,6 +129,7 @@ export const useStyles = makeStyles(
           overflow: 'hidden',
           whiteSpace: 'nowrap',
           fontWeight: theme.typography.fontWeightMedium,
+          textTransform: 'capitalize',
```
[the remark this answers is not in the document]
Comment (dtassone) [reply packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:132]: Yes Happy to remove. It was just to make the title nicer.

File: packages/storybook/src/stories/grid-rows.stories.tsx [hunk 2]
```diff
@@ -586,3 +588,271 @@ export function EditCellSnap() {
...
+          param.field
+        },
+                        triggered by ${event!.type}
+      `);
+      },
+    );
+  }, [apiRef]);
+
+  React.useEffect(() => {
+    return apiRef.current.subscribeEvent('cellExitEdit', () => {
```
[the remark this answers is not in the document]
Comment (dtassone) [reply packages/storybook/src/stories/grid-rows.stories.tsx:844]: enter/exit or start/end? 🤔

[the remark this answers is not in the document]
Comment (oliviertassinari) [reply packages/storybook/src/stories/grid-rows.stories.tsx:844]: I have used enter/exist in the initial RFC only for matching the terminology of ag-Gird, to reduce the learning curve for me developers switching. No real preference.

File: packages/grid/_modules_/grid/components/cell/GridCell.tsx [hunk 3]
```diff
@@ -66,13 +67,25 @@ export const GridCell: React.FC<GridCellProps> = React.memo((props) => {
     },
   );
 
+  const publishBlur = React.useCallback(
+    (eventName: string) => (event: React.SyntheticEvent) => {
+      const params = apiRef!.current.getCellParams(rowId, field || '');
+      // We don't trigger blur when the focus is on an element in the cell.
+      if (event.target === params.element) {
```
[the remark this answers is not in the document]
Comment (dtassone) [reply packages/grid/_modules_/grid/components/cell/GridCell.tsx:74]: `Params.element` is the cell element even when it is triggered from an element within it as it comes from params. So I believe here both should work.

[the remark this answers is not in the document]
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/cell/GridCell.tsx:74]: `params.element === event.currentTarget`

**Review by oliviertassinari — APPROVED** [review]

**Pull request merged**